src/main/java/org/sonarlint/intellij/editor/SonarExternalAnnotator.java

```diff
@@ -0,0 +1,176 @@
+    Collection<IssueStore.StoredIssue> issues = annotationResult.store.getForFile(file);
+    for (IssueStore.StoredIssue i : issues) {
+      // reject non-null ranges that are no longer valid. It probably means that they were deleted from the file.
+      if (i.range() == null || i.range().isValid()) {
+        addAnnotation(i, holder);
+      }
+    }
+  }
+
+  @Override
+  public AnnotationContext collectInformation(@NotNull PsiFile file) {
+    Project project = file.getProject();
+    IssueStore store = project.getComponent(IssueStore.class);
+    return new AnnotationContext(store);
+  }
+
+  @Override
+  @Nullable
+  public AnnotationContext doAnnotate(AnnotationContext collectedInfo) {
+    return collectedInfo;
+  }
+
+  private void addAnnotation(IssueStore.StoredIssue i, AnnotationHolder annotationHolder) {
+    Issue issue = i.issue();
+    PsiElement el = i.element();
+    TextRange textRange;
+
+    // calculate range either from the MarkerRange or from the element itself
+    if (i.range() != null) {
+      textRange = createTextRange(i.range());
+    } else {
+      textRange = i.element().getTextRange();
+    }
+
+    String msg = getMessage(issue);
+    String htmlMsg = getHtmlMessage(issue);
+
+    Annotation annotation = annotationHolder.createAnnotation(getSeverity(issue.getSeverity()), textRange, msg, htmlMsg);
+    if (i.range() == null && el instanceof PsiFile) {
+      // It is a file issue if it has no range within a file-type element
+      annotation.setFileLevelAnnotation(true);
+    } else {
+      annotation.setEnforcedTextAttributes(getTextAttrs());
+    }
+
+    annotation.setGutterIconRenderer(new SonarGutterIconRenderer(msg, issue.getRuleKey(), issue.getRuleKey()));
+  }
+
+  private TextAttributes getTextAttrs() {
+    Color c;
+    if(!unitTest) {
+      c = JBColor.YELLOW.darker();
+    } else {
+      c = Color.yellow;
+    }
+    return new TextAttributes(null, null, c, EffectType.WAVE_UNDERSCORE, Font.PLAIN);
+  }
+
+  /**
+   * Check in IntelliJ {@link com.intellij.codeInsight.daemon.impl.LocalInspectionsPass#createHighlightInfo} and
+   * {@link InspectionDescriptionLinkHandler}
+   */
+  private String getHtmlMessage(Issue issue) {
+    @NonNls
+    final String link = " <a "
+      + "href=\"#sonarissue/" + issue.getRuleKey() + "\""
+      + (isDark() ? " color=\"7AB4C9\" " : "")
+      + ">" + issue.getRuleKey()
+      + "</a> ";
+    return XmlStringUtil.wrapInHtml(link + XmlStringUtil.escapeString(issue.getMessage()));
+  }
+
+  private boolean isDark() {
+    // in unit tests on a system without gui (like travis) UIUtil will fail, and Application is not available in simple tests
+    return unitTest || ApplicationManager.getApplication().isUnitTestMode() ||
+      UIUtil.isUnderDarcula();
+  }
+
+  private static String getMessage(Issue issue) {
+    return issue.getRuleKey() + " " + issue.getMessage();
+  }
+
+  private static HighlightSeverity getSeverity(@Nullable String severity) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "getSeverity" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)